scipy/optimize/_milp.py

```diff
@@ -0,0 +1,359 @@
+        `bounds` or take value ``0``.
+
+        ``3`` : Semi-integer variable; decision variable must be an integer
+        within `bounds` or take value ``0``.
+
+        By default, all variables are continuous. `integrality` is converted
+        to an array of integers before the problem is solved.
+    bounds : scipy.optimize.Bounds, optional
+        Bounds on the decision variables. Lower and upper bounds are converted
+        to double precision arrays before the problem is solved. The
+        ``keep_feasible`` parameter of the `Bounds` object is ignored. If
+        not specified, all decision variables are constrained to be
+        non-negative.
+    constraints : sequence of scipy.optimize.LinearConstraint, optional
+        Linear constraints of the optimization problem. Arguments may be
+        one of the following:
+
+        1. A single `LinearConstraint` object
+        2. A single tuple that can be converted to a `LinearConstraint` object
+           as ``LinearConstraint(*constraints)``
+        3. A sequence composed entirely of objects of type 1. and 2.
+
+        Before the problem is solved, all values are converted to double
+        precision, and the matrices of constraint coefficients are converted to
+        instances of `scipy.sparse.csc_array`. The ``keep_feasible`` parameter
+        of `LinearConstraint` objects is ignored.
+    options : dict, optional
+        A dictionary of solver options. The following keys
+        are recognized.
+
+        disp : bool (default: ``False``)
+            Set to ``True`` if indicators of optimization status are to be
+            printed to the console during optimization.
+        presolve : bool (default: ``True``)
+            Presolve attempts to identify trivial infeasibilities,
+            identify trivial unboundedness, and simplify the problem before
+            sending it to the main solver. It is generally recommended
+            to keep the default setting ``True``; set to ``False`` if
+            presolve is to be disabled.
+        time_limit : float
+            The maximum time in seconds allotted to solve the problem;
+            default is the largest possible value for a ``double`` on the
+            platform.
+
+    Returns
+    -------
+    res : OptimizeResult
+        An instance of :class:`scipy.optimize.OptimizeResult`. The object
+        is guaranteed to have the following attributes.
+
+        status : int
+            An integer representing the exit status of the algorithm.
+
+            ``0`` : Optimal solution found.
+
+            ``1`` : Iteration or time limit reached.
+
+            ``2`` : Problem is infeasible.
+
+            ``3`` : Problem is unbounded.
+
+            ``4`` : Other; see message for details.
+
+        success : bool
+            True when an optimal solution is found, the problem is determined
+            to be infeasible, or the problem is determined to be unbounded.
```

Comment:
Success here is intended to refer to whether the _algorithm terminated successfully_ or not. Under that definition, determining that the problem is infeasible or unbounded is would be considered successful.

We could reconsider this, as apparently that's not how `linprog` always defines success. There, infeasible and unbounded have `success: False`. I forgot.

In that case, we might want to remove `success` entirely. I don't see a need for the result object to give its opinion on what it considers to be successful.

Reply:
What do you think about removing `success` @tupui @mckib2?

Reply:
Oh right, I re-read this and now understand what you meant. With the phrasing I got confused as it was not clear to me that True applied to the 3 following cases.

I would leave the attribute, since it's coherent with the other optimizers.

Reply:
That's the problem, actually. As it is written here, it's _not_ consistent with the other optimizers.

> apparently that's not how `linprog` always defines success. There, infeasible and unbounded have `success: False`. I forgot.

So as written here in `milp`: infeasible, unbounded, and optimal are all `success: True`.
In `linprog`: only optimal is `success: True`.

I don't really agree with the distinction for success used in `linprog`. It's not much more convenient for the user to test for `success == True` than `status == 0`. Having `success` as an attribute there is really expressing an opinion more than adding new information.

As written in `milp`, it adds information about which error codes are the result of completely successful termination of the algorithm. But on the other hand, which error codes correspond with success could just be in the documentation.

So what do we do? Make `success` consistent with `linprog`, make `success` more useful than it is in `linprog`, or remove it entirely?

Reply:
> So what do we do? Make `success` consistent with `linprog`, make `success` more useful than it is in `linprog` (case we do here), or remove it entirely?

+1 for removing.  It does mean consumers of the results objects might need to write a little bit more code to figure out the status of the result, but I don't think this is a bad thing.  If we do keep it, I'm +0.5 for reverting to `linprog` behavior here with the thought that if you're going to providing similar results interfaces, they should behave the same

Reply:
With that in mind, what do you think @tupui?
@Kai-Striega we're talking about leaving `success` out of the `OptimizeResult` for the new `milp` function. Your thoughts would be appreciated!

Reply:
@Kai-Striega could we have your opinion on whether we should include `success` as a field of the `OptimizeResult` returned by `milp`? `linprog` returns `success=True` only when `status=0` (Optimal). This PR has `success=True` when the algorithm terminates successfully, whether that means the problem is optimal, unbounded, or infeasible (`status` 0, 2, or 3). We don't want to have `success` and be inconsistent with `linprog`, but I question the utility of `success` if it's equivalent to `status=0`, so we're talking about taking it out. What do you think?

Reply:
I'm +1 on staying consistent with the ``linprog`` interface. I suspect that the set of users who use ``linprog`` will be similar to the set of users using ``milp`` and it's going to be easier on them if they stay the same. Plus it is still an optimizer and I think it should provide as many of the ``OptimizeResult`` field as possible.

Reply:
OK. Looks like we're split here, but I'll let go of it so we can push the PR forward. I'll set `success = (status == 0)` in the next commit.